### PR TITLE
Ensure Forwarded and X-Forwarded values are the same

### LINKED ISSUE #35751: Reconsider the precedence of Forwarded vs X-Forwarded-For
### Description

Previously change https://github.com/quarkusio/quarkus/pull/25117 introduced the option to set both `quarkus.http.proxy.allow-forwarded` and `quarkus.http.proxy.allow-x-forwarded` at the same time. It is rightly highlighted in the docs that this combination is not recommended and there can be security implications if used ([doc link](https://quarkus.io/guides/http-reference#reverse-proxy)). 

When both are enabled, Quarkus currently uses `Forwarded` over `X-Forwarded-*` if both headers are present. However many popular proxies like NGINX and Envoy do not seem to support `Forwarded` - they do support `X-Forwarded-*`. As a result, application that enabled both will be subject to header forging (untrusted client setting `Forwarded`), unless user of the application realizes this and explicitly re-configures the proxy to overwrite `Forwarded`.  If the order Quarkus parses the headers would be the other way around (`X-Forwarded-*` over `Forwarded`) this would not be the case. Even if user missed reconfiguring the proxy and uses the defaults, they would be safe(r).

Another reasoning for the idea: often the backwards compatibility design principles say that old functionality should take precedence over new functionality.

If you consider this idea of swapping the priority something you'd take into Quarkus, I can contribute it.

### Implementation ideas

_No response_

### PR DESCRIPTION
**Pull request by sberyozkin**
Fixes #35751 

Enabling both `X-Forwarded` and `Forwarded` for the proxy address forwarding may pose some risks - it is very clearly mentioned in JavaDocs for both the `allowForwarded` and `allowXforwarded` properties and a warning is logged when both are enabled. This PR adds a hardening check to make sure that if both Forwarded and XForwarded headers are enabled on the proxy address forwarding path, then the header values must match.

I've been thinking that if the mismatch happens then an error must be reported and it should be `400`, as opposed to `500`. If it is agreed upon then the remaining puzzle for me is where to intercept non-JAX-RS `BadClientRequestException` in order to end the request with 401, which I added in this PR, since the forwarded handler delegates to `Vert.x HttpServerRequest`.

Also CC @luneo7

### CODE REVIEW

**Comment by sberyozkin** [comment]
I've figured out how to end it with 400, setting the response status and doing `delegate.end()` seems to work, though happy to improve it

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/ForwardedParser.java
Comment (gsmet) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/ForwardedParser.java:347]: Wouldn't it be a lot simpler using one record for both and using equals on the records?

Comment (sberyozkin) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/ForwardedParser.java:347]: @gsmet Some or all of there properties are set in the `allowForwarded` and `allowXForwarded` cases, I found it was easier this way to keep corresponding properties, the code for handling both cases is not quite symmetrical, For `allowxForwaded` I started with declaring directly, without any bean wrappers:

```
String xForwardedScheme = scheme;
        String xForwardedHost = host;
        int xForwardedPort = port;
        String xForwardedUri = uri;
        String xForwardedRemoteHost = null;
        int xForwardedRemotePort = -1;
```
and the same for the `allowForwarded` case, and got confused quickly as it is not clear when a given property can be null, and then when both code branches are finished, they have to be compared, so this is why I ended doing it this way.

I'm all for trying to make it neater, I'm just not sure if, without records, it would be much simpler to have a single class and a manual `equals` implementation... but yeah, I'm happy to tune it further...


What do you think ?

Comment (gsmet) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/ForwardedParser.java:347]: Ah, your problem is the immutable nature of the record and the fact you want to initialize them property per property?

Not a big fan but I suppose having a builder + creating a record is not ideal in a runtime path that could end up being hot.

In any case, I'm not sure I would go with two different containers given they are identical. You could have one container with a proper equals and use that.

But yeah, do as you think is best as it's important this code doesn't break.

Comment (sberyozkin) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/ForwardedParser.java:347]: @gsmet Yeah, I can collapse them into a single one at least

Comment (sberyozkin) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/ForwardedParser.java]: @gsmet It should look better now, thanks

**Comment by luneo7** [comment]
Will see if I can test this in our infra, we have one use case that our service is served through AWS API GW externally (which we also filter the exposed endpoints), and internally through VPN we hit the service hitting one AWS ALB... so each one adds the header differently, ALB adds the Forwarded, API GW adds X-Forwarded... so need to see in both entry points if this change would be a breaking change and would also render our service somewhat useless if we were to upgrade to a version with this change.
Thanks @sberyozkin for pinging me =)

**Comment by michalvavrik** [comment]
I don't really have time until Saturday, I'll re-check then, but please don't wait for my review. Thanks

**Comment by sberyozkin** [comment]
Thanks @luneo7 If it proves a breaking change that we can add a strict check property that you'd be able to turn off...

But the main reason I turned it to draft is that I'm not certain the solution is correct, because  right now it will cause a failure if, for ex, `Forwarded` headers impact  `scheme, host` only and `XForwarded` only `host`. 

I believe only those properties which have been impacted by both headers must be compared. For example, in the previous example, it should be checked that both  `Forwarded` and `XForwarded` set exactly the same `host`, and not trying to check the scheme.

Is it a correct assumption ?

**Comment by cescoffier** [comment]
Thanks @sberyozkin for addressing this! It makes sense. Honestly, I would fail if the values are not the same (and add a flag to disable this check).

**Comment by sberyozkin** [comment]
Hi @cescoffier 

Yes, it is `HTTP 400` right now if the values does not match.

Right, I can look at adding a property, I'd like to hear from @luneo7 as well when he gets a chance to test it in his setup

**Review by cescoffier — CHANGES_REQUESTED** [review]
LGTM - I would just add the flag as discussed.

**Comment by sberyozkin** [comment]
Hi @cescoffier @gsmet, I had this feeling when looking at the code, but I was confused by several things, by the #35751 issue,  so I've realized only now that `Forwarded` and `X-Forwarded` are not supported at the same time on `main`, in released versions.   
`X-Forwarded` is only checked in the `else` branch [here](https://github.com/quarkusio/quarkus/blob/main/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/ForwardedParser.java#L151), if `Forwarded` has not been processed above [here](https://github.com/quarkusio/quarkus/blob/main/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/ForwardedParser.java#L135).

So when the user says `allowForwarded=true`, then `allowXForwarded=true` is ignored. So it is always only `Forwarded` or `XForwarded`.
[
The test](https://github.com/quarkusio/quarkus/pull/25117/files#diff-c1f9d6e4fea5b4c9fc6576cae6edc7e35d7d9d872721ae9072ff6f6778b96628R28) which added to prove that both `Forwarded` and `X-Forwarded` can be used simply confirms that only `Forwarded` is taken into consideration.

The original issue, #25077, remains unresolved, while #35751 is basically invalid because `Forwarded` can not override `X-Forwarded`, because the latter is ignored altogether.

So I've moved it to draft again because it actually changes the `main` behavior and properly fixes #25077, applies a strict control by default where the same header values must match but allows X-Forwarded override some or all of `Forwarded` values which would actually address #35751 - essentially it disabling a strict mode indirectly changes the precedence - XForwarded headers will decide what can be overridden if Forwarded is present.

Comments are welcome, I'll reopen for review once everyone is happy with what is proposed here.

**Comment by luneo7** [comment]
@sberyozkin in our case it wouldn't work, just checked... `forwarded` header has the right user IP address, and `X-Forwarded-For` has internal IP from AWS that hit the ALB, since the path is API GW -> ALB... whereas hitting the service directly only has `X-Forwarded-For`.... so if we have this rule to match both headers, it would make our use case not work... it would be awesome to have a disable flag

**Comment by luneo7** [comment]
And yeah, when we have both headers, `forwarded` has precedence over `X-Forwarded-For`, so `forwarded` needs to be considered and `X-Forwarded-For` discarded

**Comment by sberyozkin** [comment]
Thanks @luneo7. 
So you have 2 types of requests, direct with `X-Forwarded` only, and indirect with `Forwarded` and  `X-Forwarded`  combined but `Forwarded` being preferred.
But the current solution for `Forwarded` and  `X-Forwarded`  combined does not allow `X-Forwarded` being preferred.

I think we need one more property, the precedence enum.

**Comment by sberyozkin** [comment]
Now, when both types of forwarded headers are enabled, and no strict control is enabled, the precedence decides which headers make `final` impact. For example, if Forwarded has a precedence then they are run `after` X-Forwarded are run to be able to overwrite whatever they set, and vice versa.

**Review by michalvavrik — APPROVED** [review]

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/ForwardedParser.java
Comment (cescoffier) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/ForwardedParser.java:152]: Can you add a log message here? Probably `debug`. I know we will have questions - that would help us understand what's going on.

Comment (cescoffier) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/ForwardedParser.java:152]: Ok, there is a debug statement in the `modifiedPropertiesMatch` method indicating which key does not match. So, it's all good.

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/ProxyConfig.java
Comment (cescoffier) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/ProxyConfig.java:62]: Can you add some Javadoc indicating how this is used?

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/ProxyConfig.java
Comment (cescoffier) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/ProxyConfig.java:64]: Should it be `X_FORWARED`?

**Review by cescoffier — CHANGES_REQUESTED** [review]
Really good. I just added a few very minor comments. Feel free to dismiss my review once fixed.

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/ForwardedParser.java
Comment (sberyozkin) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/ForwardedParser.java]: @cescoffier Sure, I also added some messages here as well to inform that the check is done due to the strict forwarded control

**Review by cescoffier — APPROVED** [review]

**Review by luneo7 — APPROVED** [review]

**Pull request merged**
